### PR TITLE
Initial security demo

### LINKED ISSUE #6: Add spiffe integration example
cc @sids-b 

Please find a proper assignee.

### PR DESCRIPTION
**Pull request by trilokgm**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**
PR is showcase security using SPIRE for identity and management for edge and user application components.

**Which issue(s) this PR fixes**
Fixes #6

### CODE REVIEW

File: security-demo/release/app-agent-conf/agent/dummy_root_ca.crt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+-----BEGIN CERTIFICATE-----
```
Comment (m1093782566) [inline security-demo/release/app-agent-conf/agent/dummy_root_ca.crt]: Please remove the cert file in source codes.

File: security-demo/release/app-agent-conf/agent/agent.conf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    data_dir = "./.app-data"
+    log_level = "DEBUG"
+    plugin_dir = "./app-agent-conf/agent/plugin"
+    server_address = "192.168.56.102"
+    server_port = "8081"
+    socket_path ="/tmp/app-agent.sock"
+    trust_bundle_path = "./app-agent-conf/agent/dummy_root_ca.crt"
+    trust_domain = "example.org"
+    umask = ""
+    join_token = ""
```
Comment (m1093782566) [inline security-demo/release/app-agent-conf/agent/agent.conf]: What's the meaning of join_token? Can we remove this field in config file?

File: security-demo/release/app-binaries/conf/edge.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+mqtt:
```
Comment (m1093782566) [inline security-demo/release/app-binaries/conf/edge.yaml]: Can we reuse the config file of edge-core?

Comment (trilokgm) [reply security-demo/release/app-binaries/conf/edge.yaml]: yes, once there is a release script, all the configs will be generated automatically.

**Comment by m1093782566** [comment]
And please don't upload binary file alone with source codes.

**Comment by sids-b** [comment]
> And please don't upload binary file alone with source codes.

In deploy edge.sh, can we add code to get get binaries and configs from the release package(we have conf and binaries in v0.2 release tarball) ??

**Comment by m1093782566** [comment]
Any update?

**Comment by trilokgm** [comment]
The changes are under test, will upload them shortly.

**Comment by trilokgm** [comment]
> > And please don't upload binary file alone with source codes.
> 
> In deploy edge.sh, can we add code to get get binaries and configs from the release package(we have conf and binaries in v0.2 release tarball) ??

For the demo, ghostunnel uses http connections and creates mtls tunnel to peer ghostunnel. Changes for using http client are not required in mainstream kubeedge. There will a feature proposal to handle these changes for connection management.

File: security-demo/release/app-agent-conf/agent/dummy_root_ca.crt (hunk 1 above)
Comment (trilokgm) [reply security-demo/release/app-agent-conf/agent/dummy_root_ca.crt]: These are demo certificates which will ease the usage of example. The certificates are taken from spire releases. Goal is to keep user agnostic of generation certificates for the demo.

Alternate way is to create another script for generating these certificates.

Let me know your suggestions based on the above rationale.

File: security-demo/release/app-agent-conf/agent/agent.conf (hunk 2 above)
Comment (trilokgm) [reply security-demo/release/app-agent-conf/agent/agent.conf]: Yes, we can remove these unused configs.

File: security-demo/release/app-agent-conf/agent/dummy_root_ca.crt (hunk 1 above)
Comment (m1093782566) [reply security-demo/release/app-agent-conf/agent/dummy_root_ca.crt]: Let's generate the certificates - it's sensitive to upload the cert files alone with source codes.

**Comment by m1093782566** [comment]
@trilokgm Please generate the cert files at runtime, rest looks good to me.

**Comment by trilokgm** [comment]
> @trilokgm Please generate the cert files at runtime, rest looks good to me.

Ok .

File: security-demo/release/log/README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# Log files place holder
```
Comment (m1093782566) [inline security-demo/release/log/README.md]: Is this file necessary?

Comment (trilokgm) [reply security-demo/release/log/README.md]: Removed , handled in build-release.sh script.

File: security-demo/release/app-binaries/conf/edge.yaml (hunk 3 above)
Comment (trilokgm) [reply security-demo/release/app-binaries/conf/edge.yaml]: I suggest to take this into consideration as part of script optimizations. For the initial version, there are too many script functions.

File: security-demo/cloud-stub/cmd/config/app.conf [hunk 5]
```diff
@@ -0,0 +1 @@
+placementURL = "http://127.0.0.1:20000"
```
Comment (m1093782566) [inline security-demo/cloud-stub/cmd/config/app.conf]: I don't think placement is required in kubeedge.

File: security-demo/release/app-binaries/config/app.conf [hunk 6]
```diff
@@ -0,0 +1 @@
+placementURL = "ws://127.0.0.1:20000"
```
Comment (m1093782566) [inline security-demo/release/app-binaries/config/app.conf]: Placement is not required in kubeedge, right?

**Comment by m1093782566** [comment]
Please address the outstanding comments and then squash the commits.

**Comment by m1093782566** [comment]
Just want to be clear, have you ever test the example with kubeedge(include cloud and edge) or this example is just tested with IEF?

File: security-demo/README.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+3\. In cloud node , execute deploy-cloud.sh.
+
+4\. In edge node , execute deploy-edge.sh.
+
+5\. Register an example device with cloud using following command. Please
+note, in the current version, cloud test application opens 30000 port
+for metadata creation (create pod or device) and 20000 port for
+communication with kubeedge edgehub. In cloud node, execute
+
+curl -XGET http://127.0.0.1:30000/device -H 'content-type:
```
Comment (m1093782566) [inline security-demo/README.md]: quota the command:

```
curl -XGET http://127.0.0.1:30000/device -H 'content-type: application/json' -d@&lt;SPIRE\_PATH&gt;/app-binaries/test-device.yaml
```

File: security-demo/cloud-stub/cmd/config/app.conf (hunk 5 above)
Comment (trilokgm) [reply security-demo/cloud-stub/cmd/config/app.conf]: The demo is based on v0.1. So placement mock is required. In the next version, when cloud components are integrated, all config and changes for cloud stub will be removed.

File: security-demo/release/app-binaries/config/app.conf (hunk 6 above)
Comment (trilokgm) [reply security-demo/release/app-binaries/config/app.conf]: The demo is based on v0.1. So placement mock is required. In the next version, when cloud components are integrated, all config and changes for cloud stub will be removed.

File: security-demo/README.md (hunk 7 above)
Comment (trilokgm) [reply security-demo/README.md]: Updated the command.

File: security-demo/README.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+## SPIRE-based Identity Management Reference Architecture
+
+Goal of this example is to achieve workload identity management for
+cloud and edge processes . A reference architecture is shown below. A
+list of things to-do is listed below in version information section.
+For further integration , an extension add-on will be developed
+integrated with kubeege for the same.
+
+![image](./doc/images/demo-arch.png)
```
Comment (m1093782566) [inline security-demo/README.md]: Please add the workflow explanation for the architecture picture.

File: security-demo/README.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+## Source Folders
+* release : Directory has configurations and scripts to be used for deployment of identity management infrastructure for Kubeedge.
+
+* app-agent-conf  : Configurations for spire agent interfacing with edge (event-bus) and user applications . Spire agent communicates to edge spire server.|
```
Comment (m1093782566) [inline security-demo/README.md]: Please remove the `|` at the end of this line.

File: security-demo/README.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+* conf : Configurations for spire server (cloud and edge) and spire agent (agent communicating to cloud spire server).
+
+* certs : Spiffe helper communicates with spire agents and download certificates in this folder . Ghostunnel is loaded with these certificates . For edge to user interface application , spiffe helper configuration and ghostunnel certs are present in event-bus and user-app folders respectively.
+
+* app-binaries : All binaries for edge , cloud and test applications.
```
Comment (m1093782566) [inline security-demo/README.md]: I am wondering if this folder is required?

File: security-demo/README.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+                           
+                    2) Registers and starts edge spire server.
+                           
+                    3) Registers and starts edge agent for edge to user app communication interface.
+
+* start-spiffe-helper.sh  : Based on the configuration (helper.conf) , starts spiffe-helper communication for certificate download and rotation. Ghostunnel is run using spiffe-helper.
+
+* commands.sh  : Abstracts spire cli commands.
+ 
+## Prerequisites
```
Comment (m1093782566) [inline security-demo/README.md]: The following prerequisites are common, so there is no need to add them?

File: security-demo/README.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+## How to configure
+### Cloud node configuration
+
+*Upstream CA – Cloud spire server configuration:*
+&lt;SPIRE\_PATH&gt;/conf/server/server.conf
+
+*Upstream CA Agent – Cloud node agent configuration :*
+&lt;SPIRE\_PATH&gt;/conf/agent/agent.conf
+
+Following fields might require modification based on the deployment environment
```
Comment (m1093782566) [inline security-demo/README.md]: Following are the fields of what configuration file?

File: security-demo/README.md [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+       server_port = "8081" <Cloud Spire Server port>
+
+       socket_path =/tmp/agent.sock"
+
+### Edge node configuration
+
+*Edge agent configuration to **connect to cloud spire server** :*
+&lt;SPIRE\_PATH&gt;/conf/agent/agent.conf
+
+Following fields might require modification based on the deployment environment
```
Comment (m1093782566) [inline security-demo/README.md]: Following are the fields of what configuration file?

File: security-demo/README.md [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+scripts.
+
+3\. Communication from user application and edge application using
+certificates issued by cloud spire server and edge spire server.
+
+## ToDo
+
+1\. Optimization for redundancies in configuration and scripts.
+
+2\. Automate build/release and remove binaries from the package.
```
Comment (m1093782566) [inline security-demo/README.md]: So there is still binaries in the package?

File: security-demo/README.md [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+3\. In cloud node , execute deploy-cloud.sh.
+
+4\. In edge node , execute deploy-edge.sh.
+
+5\. Register an example device with cloud using following command. Please
+note, in the current version, cloud test application opens 30000 port
+for metadata creation (create pod or device) and 20000 port for
+communication with kubeedge edgehub. In cloud node, execute
+
+    curl -XGET http://127.0.0.1:30000/device -H 'content-type:application/json' -d@/opt/spire/app-binaries/test-device.yaml
```
Comment (m1093782566) [inline security-demo/README.md]: Quotes the command please.

File: security-demo/cloud-stub/cmd/config/app.conf (hunk 5 above)
Comment (trilokgm) [reply security-demo/cloud-stub/cmd/config/app.conf]: This will be fixed when we upgrade to v0.2.

File: security-demo/README.md (hunk 10 above)
Comment (trilokgm) [reply security-demo/README.md]: The folder will contain binaries when release is built. No binaries are uploaded directly.

File: security-demo/README.md (hunk 13 above)
Comment (trilokgm) [reply security-demo/README.md]: Also , detail configuration to be added about spiffe helper.

File: security-demo/README.md [hunk 16]
```diff
@@ -0,0 +1,263 @@
+# Identity Management Using SPIRE for Kubeedge
+
+## Background
+
+Security is a paramount requirement for edge computing architecture as security breaches can make a complete organization to come to a halt (IIot) , data breach can lead to privacy issues and also control of the complete edge computing infrastructure. 
+
+The enable better security, following needs to be satisfied for kubeedge framework
```
[the remark this answers is not in the document]
Comment (trilokgm) [reply security-demo/README.md]: Updated.

File: security-demo/README.md (hunk 8 above)
Comment (trilokgm) [reply security-demo/README.md]: Updated.

File: security-demo/README.md (hunk 9 above)
Comment (trilokgm) [reply security-demo/README.md]: Updated.

File: security-demo/README.md (hunk 11 above)
Comment (trilokgm) [reply security-demo/README.md]: Prerequisites are based on spire, script (while read problems in older version bash) and package downloading capability dependencies.

File: security-demo/README.md (hunk 13 above)
Comment (trilokgm) [reply security-demo/README.md]: Updated the details for spiffe helper.

File: security-demo/README.md (hunk 14 above)
Comment (trilokgm) [reply security-demo/README.md]: No, updated the todo list.

File: security-demo/README.md (hunk 15 above)
Comment (trilokgm) [reply security-demo/README.md]: Updated.

**Comment by m1093782566** [comment]
LGTM, thanks!
